src/time-picker/TimePicker.js

```diff
@@ -0,0 +1,211 @@
+);
+
+const addZero = str => {
+  if (str.length === 4) {
+    return `${str}0`;
+  } else {
+    return str;
+  }
+};
+
+const insertColon = str => {
+  if (numberRegex.test(str)) {
+    if (str.length === 3) {
+      const position = str.length - 1;
+      const strWithColon = `${str.substr(0, position)}${separator}${str.substr(position)}`;
+      return strWithColon;
+    } else if (str.length === 4) {
+      const position = str.length - 2;
+      const strWithColon = `${str.substr(0, position)}${separator}${str.substr(position)}`;
+      return strWithColon;
+    }
+  }
+  return str;
+};
+
+const cleanSeparator = str => str.replace(symbolRegex, separator);
+const cleanInput = inputStr => addZero(insertColon(cleanSeparator(inputStr)));
+
+const getSpecificOptionList = (cleanedInput, timeFormat) => {
+  const inputArray = cleanedInput.split(separator);
+  const hours = parseInt(inputArray[0]);
+  const minutes = parseInt(inputArray[1]);
+  if (hours !== 0 && numberRegex.test(hours) && numberRegex.test(minutes)) {
+    const optionTime12 = [
+      { value: `${cleanedInput}am`, label: timeFormatter12(hours, minutes) },
+      { value: `${cleanedInput}pm`, label: timeFormatter12(hours + 12, minutes) }
+    ];
+    const optionTime24 = [{ value: cleanedInput, label: timeFormatter24(hours, minutes) }];
+
+    if (timeFormat === H12 && hours <= 12 ) {
+      return optionTime12;
+    } else if (timeFormat === H24) {
+      return optionTime24;
+    }
+  }
+  return [];
+};
+
+const Integer = t.refinement(t.Number, n => n % 1 === 0, 'Integer');
+const Hour = t.refinement(Integer, int => int >= 0 && int <= 23, 'Hour');
+const Minute = t.refinement(Integer, int => int >= 0 && int <= 59, 'Minute');
+const TimeFormat = t.enums.of([H12, H24], 'TimeFormat');
+const Time = t.struct({
+  hours: Hour,
+  minutes: Minute
+}, 'Time');
+
+const Props = t.refinement(t.struct({
+  onChange: t.Function,
+  value: t.maybe(Time),
+  minTime: t.maybe(Time),
+  maxTime: t.maybe(Time),
+  interval: t.maybe(Integer),
+  placeholder: t.maybe(t.String),
+  timeFormat: t.maybe(TimeFormat),
+  id: t.maybe(t.String),
+  className: t.maybe(t.String),
+  style: t.maybe(t.Object)
+}), ({ value, minTime, maxTime }) => (
+  lteTime(minTime, maxTime) && (!value || (lteTime(value, maxTime) && lteTime(minTime, value)))
+), 'Props');
+
+@skinnable()
+@props(Props)
+export default class TimePicker extends React.Component {
+
+  static defaultProps = {
+    placeholder: `--${separator}--`,
+    timeFormat: H24,
+    interval: 30,
+    minTime: { hours: 0, minutes: 0 },
+    maxTime: { hours: 23, minutes: 59 }
+  }
+
+  /**
+    We use filterOptions because seems that is not possible
+    to change the options from outside of this function,
+    then we use this tricky hack and we workaround this creating a new set
+    of options as we wish and filter it if needed, or just returning a brand new array
+  **/
+  createOptionsList = (_, inputStr) => {
```

Comment:
This is actually commented in the code base as well, because that function works on top of an hack we have to make, and this is because the version of react-select we are using will not support changing `options` in other ways. The main point of misunderstanding here is that it will return an array that actually can be 3 different things.

1) `undefined` case (the user just focus the input): pre generated list, with all the needed results computed using min,max,interval and everything we need.
2) The filtered list, just when the input is that format `H`, `HH`, `HH:` actually the options list will be regenerated completely and filtered.
3) In all the other valid use format like: `HH:m`,`HH:mm`, `HHm`, `HHmm` the list will be a completely new one, generated on the basis of timeFormat (12H,24H).

Refactor proposal:
- Naming `createOptionsList` dosent means nothing to me right now. Something better can be optionsManager or something like that. 
- It's wired that a single function return another blackbox function that returns arrays that you don't know how are they generated (one time filtered, one time generated, one time both)
- performance issues, is not SO expensive in terms of performance cause the arrays of options (when it need to be generated in the case 1) can have maximum a size of 48, but is just pointless to recreate it even when is not needed. That logic can be moved into a dedicated function.